proxylib/proxylib.go

```diff
@@ -0,0 +1,198 @@
+	// Find the parser for the proto
+	parserFactory := ParserFactories[proto]
+	if parserFactory == nil {
+		return C.FILTER_UNKNOWN_PARSER
+	}
+	_, port, err := net.SplitHostPort(dstAddr)
+	if err != nil {
+		return C.FILTER_INVALID_ADDRESS
+	}
+	dstPort, err := strconv.ParseUint(port, 10, 32)
+	if err != nil || dstPort == 0 {
+		return C.FILTER_INVALID_ADDRESS
+	}
+	connection := &Connection{
+		Id:         connectionId,
+		Ingress:    ingress,
+		SrcId:      srcId,
+		DstId:      dstId,
+		SrcAddr:    srcAddr,
+		DstAddr:    dstAddr,
+		Port:       uint32(dstPort),
+		PolicyName: policyName,
+		Orig:       Direction{InjectBuf: origBuf},
+		Reply:      Direction{InjectBuf: replyBuf},
+	}
+	connection.Parser = parserFactory.Create(connection)
+	if connection.Parser == nil {
+		// Parser rejected the new connection based on the connection metadata
+		return C.FILTER_POLICY_DROP
+	}
+
+	mutex.Lock()
+	connections[connectionId] = connection
+	mutex.Unlock()
+
+	return C.FILTER_OK
+}
+
+// Each connection is assumed to be called from a single thread, so accessing connection metadata
+// does not need protection.
+//
+// OnData gets all the unparsed data the datapath has received so far. The data is provided to the parser
+// associated with the connection, and the parser is expected to find if the data frame contains enough data
+// to make a PASS/DROP decision for the whole data frame. Note that the whole data frame need not be received,
+// if the decision including the length of the data frame in bytes can be determined based on the beginning of
+// the data frame only (e.g., headers including the length of the data frame). The parser returns a decision
+// with the number of bytes on which the decision applies. If more data is available, then the parser will be
+// called again with the remaining data. Parser needs to return MORE if a decision can't be made with
+// the available data, including the minimum number of additional bytes that is needed before the parser is
+// called again.
+//
+// The parser can also inject at arbitrary points in the data stream. This is indecated by an INJECT operation
+// with the number of bytes to be injected. The actual bytes to be injected are provided via an Inject()
+// callback prior to returning the INJECT operation. The Inject() callback operates on a limited size buffer
+// provided by the datapath, and multiple INJECT operations may be needed to inject large amounts of data.
+// Since we get the data on one direction at a time, any frames to be injected in the reverse direction
+// are placed in the reverse direction buffer, from where the datapath injects the data before calling
+// us again for the reverse direction input.
+//
+//export OnData
+func OnData(connectionId uint64, reply, endStream bool, data *[]string, filterOps *[]C.FilterOp) C.FilterResult {
```

Comment:
this came from the fact that I thought you had
```
op, bytes := connection.Parser.OnData(reply, endStream, (*data)[uint:], offset)
```
instead of
```
op, bytes := connection.Parser.OnData(reply, endStream, (*data)[unit:], offset)
```

(check the `uint` vs `unit`)